sys/arm64/arm64/copyinout.S

```diff
@@ -224,3 +224,128 @@ ending:
 	mov	x0, xzr		/* return 0 */
 	ret
 	.size	copycommon, . - copycommon
+
+#if __has_feature(capabilities)
+/*
+ * Copy specified amount of data from the user to kernel space, preserving
+ * capability tags
+ *
+ * int copyincap(const void * __capabilityuaddr, void *kdaddr, size_t len)
+ */
+ENTRY(copyincap)
+	cbz	x2, 1f
+	adds	x3, x0, x2
+	b.cs 	copyio_fault_nopcb
+	ldr	x4, =VM_MAXUSER_ADDRESS
+	cmp	x3, x4
+	b.hi	copyio_fault_nopcb
+
+	/* Check the length is greater than 1 capability */
+	cmp	x2, #16
+	b.lo	copycommon
+	/* And the addresses will both align on a capability */
+	eor	x3, x0, x1		/* Compare the pointers */
+	and	x3, x3, #0xf		/* Check the lower bits */
+	cbnz	x3, copycommon		/* If different don't copy tags */
+
+	b	copycapcommon
+
+1:	mov	x0, xzr		/* return 0 */
+	ret
+END(copyincap)
+
+/*
+ * Copy specified amount of data from kernel to the user space, preserving
+ * capability tags
+ *
+ * int copyoutcap(const void *kaddr, void * __capability udaddr, size_t len)
+ */
+ENTRY(copyoutcap)
+	cbz	x2, 1f
+	adds	x3, x1, x2
+	b.cs 	copyio_fault_nopcb
+	ldr	x4, =VM_MAXUSER_ADDRESS
+	cmp	x3, x4
+	b.hi	copyio_fault_nopcb
+
+	/* Check the length is greater than 1 capability */
+	cmp	x2, #16
+	b.lo	copycommon
+	/* And the addresses will both align on a capability */
+	eor	x3, x0, x1		/* Compare the pointers */
+	and	x3, x3, #0xf		/* Check the lower bits */
+	cbnz	x3, copycommon		/* If different don't copy tags */
+
+	b	copycapcommon
```

Comment:
```suggestion
	b	copycapcommon
	cvtp	c0, x0
```
This might be sufficient to call copycapcommon with capability args?

Reply:
No, I don't want to be merging known-broken code. If it's not ready, do it in a later PR.

Reply:
well copycapcommon obviously also needs adjusting then. But isn't that exactly what we do for RISC-V? We derive from DDC, so I don't see how this is different? Although it sems `cvtp` derives from pcc so it should be `CVTD`.

Reply:
Kernel pointers are passed in as addresses and derived from DDC to give capabilities, but user pointers are passed in as capabilities and used as-is. Then `copycapcommon` always deals with capabilities.

Reply:
Yes that's exactly what I was suggesting, I guess this was confusing due to my comment earlier.